File: vol1/ch6/LeetCode2415.cpp
class Solution {
public:
    TreeNode* reverseOddLevels(TreeNode* root) {
        if(root->left==NULL && root->right==NULL)
            return root; 
        queue<TreeNode*> qu;
        qu.push(root);                              //根结点进队
        int curl=0;
        while (!qu.empty()) {                           //队不空循环
            int n=qu.size();
            vector<TreeNode*> curp;
            for(int i=0;i<n;i++) {
                TreeNode*p=qu.front(); qu.pop();           //出队一个结点p
                curp.push_back(p);
                if (p->left!=NULL)                      //有左孩子时将其进队
                    qu.push(p->left);
                if (p->right!=NULL)                     //有右孩子时将其进队
                    qu.push(p->right);
            }
            if(curl%2==1) rev(curp);
            curl++;
        }
        return root;
    }

    void rev(vector<TreeNode*> &a) {			//反转a
        int i=0,j=a.size()-1;
        while(i<j) {
            swap(a[i]->val,a[j]->val);
            i++; j--;
        }
    }
};

File: vol1/ch6/LeetCode2415.py
from collections import deque
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root.left==None and root.right==None:
            return root 
        qu=deque()      #定义一个队列qu
        qu.append(root)              #根结点(层次为1)进队
        curl=0
        while qu:                    #队不空循环
            n=len(qu)
            curp=[]
            for i in range(0,n):
                p=qu.popleft()           #出队一个结点
                curp.append(p)                 #当前层的结点值添加到curp
                if p.left!=None:                      #有左孩子时将其进队
                    qu.append(p.left)
                if p.right!=None:                     #有右孩子时将其进队
                    qu.append(p.right)
            if curl%2==1: self.rev(curp)
            curl+=1
        return root

    def rev(self,a):							#反转a
        i,j=0,len(a)-1
        while i<j:
            tmp=a[i].val					#注意应该采用传统交换方式
            a[i].val=a[j].val
            a[j].val=tmp
            i+=1; j-=1
